### ISSUE TITLE
./autogen.sh --prefix=/usr fails on systems without gnome-shell

### ISSUE THREAD

**Issue by smurphos** [jnsh/arc-theme#1]
# READ CAREFULLY BEFORE SUBMITTING

**This list is for avoiding duplicated issues, if your issue isn't listed, you should remove this and continue.**

- The theme doesn't work in a distribution that have a GTK+ 3 version lower than 3.14.
  - Arc only supports versions since 3.14, you should update to latest versions.
- The theme is broken after a major GTK+ 3 update (i.e. 3.18 to 3.20 or 3.20 to 3.22).
  - Update the package or install from sources.
- The theme is broken after installing from sources.
  - Possibly a bug on GTK+ 3 version checking, use `--with-gnome=<version>` flag with `./autogen.sh`.
- Theme isn't transparent on GNOME applications without CSD (i.e. Ubuntu's applications).
  - Transparency isn't supported for non-CSD applications.
- Theme doesn't build from AUR in Arch/Manjaro and others.
  - Ensure you have the `base-devel` group installed.

### Checks

*Ensure you...*

- **Really** read the list above and checked if your issue is or not duplicated.
- **Really** read the README and there is nothing about your issue.
- **Really** checked if the theme is installed or built specifically for the GTK+ 3/distribution version you're using.

### Details

- Distribution (with version): Mint 19.1 
  - Base distribution (if have, with version): Ubuntu 18.04
- Desktop environment (with version): Cinnamon 4.0.10
- Arc version/commit (don't put "latest"): https://github.com/jnsh/arc-theme/commit/7bb1d430f5bc4cf29286d1e5e76b9558c871a666
- GTK+ 3 version: 3.22.30
  - Murrine engine version: 0.98.2-2ubuntu1
  - `gnome-themes-standard` version: 3.28-1ubuntu1
- Installation method (package or source): Source

Hi jnsh, I'm an occasional contributor over at NicoHood's fork (mainly to the Cinnamon theme) and was intrigued by your fork and particularly the Cinnamon versioning so I thought I'd give it a test run.

This issue is simply re the process of building from source. On my system with no gnome shell installed the standard `./autogen.sh --prefix=/usr` falls over when it fails to locate gnome-shell.

`./autogen.sh --prefix=/usr --disable-gnome-shell` works fine.

I can't recall if it's an issue in NicoHood's fork as-well. Will test.

```
./autogen.sh --prefix=/usr
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: configure.ac: creating directory build-aux
autoreconf: configure.ac: not using Libtool
autoreconf: running: /usr/bin/autoconf --force
autoreconf: configure.ac: not using Autoheader
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:12: installing 'build-aux/install-sh'
configure.ac:12: installing 'build-aux/missing'
autoreconf: Leaving directory `.'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether ln -s works... yes
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
Detected GTK3 3.22.30
Building GTK3 theme 3.20
checking for gnome-shell... no
        
configure: error: Could not find gnome-shell from $PATH.
configure: WARNING: cache variable ac_cv_prog_GNOME_SHELL_FOUND contains a newline

```

**Comment by jnsh** [comment]
Thank you for the report.

That's how the build works by design. The build system needs to know the version for gnome-shell theme (and other versioned themes i.e. cinnamon and gtk3) to be built, and if that isn't specified with `--with-gnome-shell=<version>`, it tries to determine it from `gnome-shell` binary. If that all fails, the build fails as well to avoid issues resulting from invalid theme version.

However, the error message could definitely be improved to avoid confusion, and I added this to my TODO. If you have other suggestions for improving the situation, let me know.




Aside from that, if you notice any issues or anything that could be improved with the Cinnamon theme, or anything else, please let me know. I'm using gnome as my desktop, and it's possible I've missed even some obvious issues with other desktop environments, even though I try to test everything thoroughly. There could also be something that's already fixed in the NicoHood fork, since I haven't had time to go trough all the recent cinnamon related commits over there.

Also, please keep in mind, that while this fork is perfectly usable and shouldn't be in worse shape than the NicoHood fork, I'm still not considering it "stable" and ready for the masses. It's also still a bit unsure whether I will only keep developing it as a separate fork, or if we'll merge our projects with the NicoHood fork. I do appreciate all testing in any case :)

**Comment by smurphos** [comment]
No worries - I haven't tested thoroughly yet but spotted a few cosmetic issues on the 4.0 version on an initial quick test (text misalignment in the calendar and no way to distinguish the focused app in the grouped window-list where two that stood out).  When I've got a bit of time I'll have a proper run through and report back.

Regarding the versioning I am intrigued as to how you'd see this working in a deb packaged version of the theme. I guess the deb could have an installation script that checks the Cinnamon version and builds from the correct sccs on the fly? AFAIK Cinnamon doesn't support versioning in an installed theme in the same way as Gnome-Shell.

**Comment by jnsh** [comment]
> I haven't tested thoroughly yet but spotted a few cosmetic issues on the 4.0 version on an initial quick test (text misalignment in the calendar and no way to distinguish the focused app in the grouped window-list where two that stood out).

Thanks for these, I can confirm the issue with grouped-window-list. I remember having some issues while working on it, that I thought may have been upstream issue, but I'll have another look at it.

I don't notice anything wrong with the text on the calendar though (if you mean the popup that opens when clicking the clock on the panel). If you can open a new issue for that with some more details when you get a chance, that would be great.

The versioning for cinnamon is not any different from how the gnome-shell and gtk3 versioning works. The version is determined either from build options, or from packages installed on the build environment, and the theme is built from the scss for that version. I think gtk3 supports picking the correct theme on runtime from multiple installed theme versions, but the build system for Arc has always installed only single version. For gnome-shell and cinnamon, there's no run-time versioning for the theme that I'm aware of.

AFAIU, .deb packages are usually built for specific distribution and release, and the theme versions have to be specified correctly during the packaging. I don't think that the addition of cinnamon versioning shouldn't cause difficulties with that, since gnome-shell and gtk3 themes were already built in the same manner.

The main purpose of versioning cinnamon and gnome-shell themes is simplifying the development process, since you don't have to keep supporting and testing for older versions.

**Comment by smurphos** [comment]
The issue with the deb packaging and cinnamon versioning that may require some thought to ensure the correct theme is installed is that for example a deb targeting bionic could end up on a system with Cinnamon 3.6.x (Ubuntu), 3.8.x (Mint 19), 4.0.x (Mint 19.1), and eventually 4.2.x and 4.4.x (forthcoming Mint versions 19.2 and 19.3).

A deb targeting Debian stretch might end up on a system with Cinnamon 3.2.x  or 4.0.x (LMDE)

GTK supports versioning at the installed theme level - e,g, you can have gtk-3.0, a gtk-3.20, a gtk-3.22 and a gtk-3.24 folders in an installed theme and gtk will use the correct version. The same doesn't apply AFAIK to gnome-shell and cinnamon, but it's not really a problem for gnome-shell in Ubuntu, as a default bionic installation will always be on a 3.28.x version.

It's certainly something worth exploring, as there are some compromises involved in maintaining a single version  cinnamon theme that can look good across the current version and former versions that are still out in the wild.

**Comment by jnsh** [comment]
> for example a deb targeting bionic could end up on a system with Cinnamon 3.6.x (Ubuntu), 3.8.x (Mint 19), 4.0.x (Mint 19.1), and eventually 4.2.x and 4.4.x (forthcoming Mint versions 19.2 and 19.3).


> A deb targeting Debian stretch might end up on a system with Cinnamon 3.2.x or 4.0.x (LMDE)

Yes, every distribution/release ships different base package versions, and other packages has to be built against those packages separately for each distribution/release. This is basic packaging stuff, and I don't see how the addition of cinnamon versioning would result in any kind of special case.


> GTK supports versioning at the installed theme level


> I did think gnome-shell did support this but apparently not?

I checked the source for user-theme shell extension, and it didn't seem to so. I'd be happy to be mistaken though, as the ability to ship all theme versions simultaneously would make shipping of the theme much simpler.

**Comment by jnsh** [comment]
I improved the error output with 45783e220e3b47bd52c2fef091043fd4caae8f5b, and I think that should address this issue.

@smurphos I went trough your Cinnamon updates from NicoHood repo and picked up the fixes I hadn't implemented earlier, apart from few I wasn't sure about from design aspect. Those fixed the calendar text alignment, among some other stuff. For the changes to window-lists, and panel-launchers I'm still thinking for the best solution, but will fix them eventually.

Thanks again for this report, and the Cinnamon testing and fixes.
